### ISSUE TITLE
Is this a functionality that is possible?

### ISSUE THREAD

**Issue by ziofil** [tensorflow/tensorflow#41836]
I have created an operation that has a custom gradient. Then I use it to define a new Keras Model, out of layers:

```python
import numpy as np
import numba
import tensorflow as tf

@numba.jit(nopython = True)
def func(param, input):
    return param*input**2

@numba.jit(nopython = True)
def gradfunc(param, input):
    return input**2

@tf.custom_gradient
def func_tf(param, input):
    def grad(dy):
        return tf.numpy_function(gradfunc, (param.numpy(), input.numpy()), tf.float32), 2*param*input
    return tf.numpy_function(func, (param.numpy(), input.numpy()), tf.float32), grad

class myLayer(tf.keras.layers.Layer):
    def __init__(self):
        super().__init__()
        
    def build(self, input_shape):
        self.param = self.add_weight("param")
        
    def call(self, input):
        return func_tf(self.param, input)
    
class myModel(tf.keras.Model):
    def __init__(self, num_layers):
        super().__init__(name='')
        self._layers = [myLayer() for _ in range(num_layers)]
        
    def call(self, input_tensor):
        for layer in self._layers:
            input_tensor = layer(input_tensor)
        return input_tensor
    
model = myModel(3)
print(model(1.5)) # <-- this works
```
This, however fails:
```python
def loss(target, output):
    tf.abs(tf.reduce_sum(target - output))**2

model.compile(
    optimizer=tf.keras.optimizers.Adam(),
    loss=loss,
    metrics=[loss])

model.fit([0.1], [0.4], batch_size=None)
```

Because `model.fit` uses `@tf.function` under the hood, so the calls to `.numpy()` in `func` and `gradfunc` are not possible (see #40508)

I have found [this answer](https://stackoverflow.com/questions/41132633/can-numba-be-used-with-tensorflow) on SO, but I think that works only if `param` were not an input to `func` and `gradfunc`.

How do I make it work?

**Comment by gowthamkpr** [comment]
@ziofil 

This is not yet possible.

But you have two choices:

a) Use Tensorflow ufuncs for your loss function, for instance not using .numpy() and replacing np.sum() by tf.reduce_sum()

b) Use NumPy ufuncs, but train eagerly, by passing run_eagerly=True in model.compile()
